### ISSUE TITLE
Primary scoreboard enhancements

### ISSUE THREAD

**Issue by jawaad-ahmad** [brata-hsdc/brata.masterserver#7]
Enhance the legacy MS as follows:

I thought the leader board was great.  Make it more readable, and possibly more entertaining.  It is a focal point of the competition.

Also factor in the following older issues into the enhancements:
- brata-hsdc/brata#57 - Ensure leader board fits all team rows on a single screen
- brata-hsdc/brata#58 - Ensure differentiation between trying a challenge and failing (zero score) vs. not attempting a challenge at all yet (black dash on the old leader board)
  - Thought: Initialize scores to -1 and only total non-negative scores
- brata-hsdc/brata#61 - Live update leader board using AJAX
- brata-hsdc/brata#65 - Enhance leader board icons
  - Reference issue for details
- brata-hsdc/brata#66 - Rows difficult to distinguish from neighboring rows

Let's try to get some time from someone in Human Factors to review mock-ups before we implement this.

(I think anything front-end should be reviewed with mock-ups before we actually start coding it. If anything changes on paper, it's a lot less effort than reworking code.)

This breaks up Issue brata-hsdc/brata#74 Part 4. Issue brata-hsdc/brata#74 was later continued as Issue #4.

**Comment by ellerychan** [comment]
If we do proper separation of content and styling, we will be able to proceed in parallel.  Let's make good use of CSS and Django's templates, and we will have a lot of flexibility with layout and styling.

I would love to see some mock-ups early on.

**Comment by jawaad-ahmad** [comment]
Agreed. What's recommended for content vs. styling these days? Do we do MVVM-style separation using Knockout? Also I updated the title to clarify this is different from the mobile scoreboard.

**Comment by ellerychan** [comment]
Never heard of _MVVM_ before, but I just looked at it.  It looks like MVC with C renamed to VM.

Again I'm going to recommend not adding another package unless the overall complexity goes down, not up.  I was hoping we could get by with just **jquery** to simplify the JavaScript.  But it is good to know that things like **[Knockout](http://knockoutjs.com/)** exist in case we need them.

**Comment by DragonmasterJ** [comment]
That' correct, I've heard scholars says it's more MVP, but that doesn't make a difference for us. 

I hate to be the guy that's behind again, but I'm assuming the JavaScript ties the Django app to the CSS? Is that it's own separate layer to worry about, or is it more like Java where it'll more tied into the view than being a separate entity?

**Comment by ellerychan** [comment]
CSS properties are associated with the HTML elements in the DOM by an element's _class_ attribute, _id_ attribute, element type (`a`, `div`, `ul`, etc.), or nesting relationship (`p` elements that immediately follow an `h1`, for example).  JavaScript is not needed.  However, if JavaScript is used, it can set those attributes dynamically.

**Comment by jawaad-ahmad** [comment]
One item briefly brought up was 3D graphics and status displays. I assume it pertained to this.

There are several fixes noted above for the existing leaderboard. That leaderboard seems like something that can continue for years if maintained, whereas my understanding of the proposed 3D display would only be good for this year.

As such, the changes above look like they're pretty small. Should we implement those fixes now under this issue, and then open a lower-priority issue for the fancy leader board to work after we finish all the higher-priority Master Server issues?

**Comment by ellerychan** [comment]
I think we should split the 3D displays out.  I created another issue for them (#15).

> my understanding of the proposed 3D display would only be good for this year

I think the 3D graphics display should be thought of as just a graphical status display, driven by an external data source.  It could be 3D or it could be simple 2D graphics.  It could evolve into something highly configurable, so I think it could easily have life beyond this year's event. 

> the changes above look like they're pretty small. Should we implement those fixes now under this issue?

I think we can handle them in this issue, although I think some of them will be OBE with the new MS.  But we should leave this issue to make sure they are handled in the new implementation.

**Comment by jawaad-ahmad** [comment]
I believe a bunch of the original issues listed here should be addressed with the current changes to the scoreboard. I'm taking this so I can keep these in mind while working the updates to make sure the original issues get addressed.
